src/components/Rating/Rating.tsx

```diff
@@ -0,0 +1,110 @@
+import { useState } from "react";
+import "./rating.scss"; // Assuming you have a Sass file for styling
+
+
+export interface RatingProps {
+    totalStars: number;
+    shape?: string;
+    disabled?: boolean
+    readOnly?: boolean;
+    readOnlyValue?: number
+}
+
+const Rating: React.FC<RatingProps> = ({
+    totalStars,
+    shape = '\u2605',
+    disabled = false,
+    readOnly = false,
+    readOnlyValue = 5
+}) => {
+    const defaultRating = readOnly ? readOnlyValue : 0;
+
+    const numberOfStars = totalStars
+    const [rating, setRating] = useState(defaultRating);
+    const [hoveredRating, setHoveredRating] = useState(0);
+
+    const handleStarClick = () => {
+        if (!disabled && !readOnly) {
+            setRating(hoveredRating);
+
+        }
+    };
+
+    const handleStarHover = (e: React.MouseEvent<HTMLSpanElement>, index: number) => {
+        if (!disabled && !readOnly) {
+            const starWidth = (e.target as HTMLSpanElement).offsetWidth;
+            const mouseX = e.clientX - (e.target as HTMLSpanElement).getBoundingClientRect().left;
+            const hoverWidth = (mouseX / starWidth) * 100;
+
+            // Round hoverWidth to nearest 10 for more predictable behavior
+            const roundedHoverWidth = Math.round(hoverWidth / 10) * 10;
+            // Determine hoveredRating based on hoverWidth
+            if (roundedHoverWidth < 25) {
+                setHoveredRating(index + 0.5);
+            } else if (roundedHoverWidth >= 25 && roundedHoverWidth < 75) {
+                setHoveredRating(index + 1);
+            } else {
+                setHoveredRating(index + 1.5);
+            }
+        }
+    };
+    const handleStarLeave = () => {
+        if (!disabled) {
+            setHoveredRating(0);
+        }
+    };
+
+    return (
+        <>
+            <h1> Rating {rating}</h1>
+            <div className="container">
+                {/* Actual rating stars */}
+                <div className="realstarsdiv">
```

Comment:
Why duplicate these? I might suggest using the same set of stars to display the currently-hovered and stored rating - while the mouse is hovered, display the currently-hovered rating, and when it leaves, then display the stored rating.

As a bonus, doing this would fix what I find to be a slightly unintuitive behavior - when a rating is set higher than the currently-hovered rating, the stored rating appears to be displayed and there's no visual feedback indicating my currently-hovered rating (except for a slight little difference in the border that looks more unintentional than not.)


https://github.com/user-attachments/assets/8ab3ac65-dc3e-4f8b-9903-839b2800e649

Reply:
The duplication made it easier for me to display the hovered vs actual selected rating. Ill try to see if i can make this less repetitive though.  I wanted to keep track of the hovered rating without changing already selected rating. I think the hovered rating not displaying if its lower than the current selected rating is normal behavior for this kind of component. It mimics the behavior from other component libraries (example https://daisyui.com/components/rating/ )